### PR TITLE
Added "Select All" feature to Delete Unused Branches plugin

### LINKED ISSUE #3522: Add Select All button to the Delete obsolete branches form
It would be useful to have a `Select all` button on the Delete obsolete branches form in order to quickly mark all listed branches for deletion.

### PR DESCRIPTION
**Pull request by oriash93**
Fixes #3522

Changes proposed in this pull request:
- Added "Select All" feature to Delete Unused Branches plugin
 
Screenshots before and after (if PR changes UI):
- Before:
image

- After:
image

What did I do to test the code and ensure quality:
- Pressed the button and made sure it selects all branches.

Has been tested on (remove any that don't apply):
- GIT 2.10 and above
- Windows 10

### CODE REVIEW

**Comment by vbjay** [comment]
Yes.

**Comment by drewnoakes** [comment]
I'd say either add a "Select none" button alongside, or adding a three-state checkbox at the bottom, aligned with the column of checkboxes above.

**Comment by RussKie** [comment]
I like "_adding a three-state checkbox at the bottom, aligned with the column of checkboxes above_" option, though I maybe add it above the grid...

**Comment by oriash93** [comment]
Eventually went with @RussKie's idea, a checkbox in the column header that selects/deselects all the checkboxes.
However, I had to create a `DataGridViewCheckBoxHeaderCell` class that inherits from `DataGridViewColumnHeaderCell` for this. If anyone finds a better way to do this, be my guest.

**Comment by vbjay** [comment]
That is the right way to do it.

**Comment by vbjay** [comment]
> Similar solution would be nice for "Push multiple branches" too.

Yes.  I would ignore the push with force column if this is implemented.  I would want them to click each checkbox that is to be forced manually.  A check all for that column is not a good idea.

**Comment by oriash93** [comment]
Nice idea. I can take it up if you want to open a separate issue.
Regardless, let this PR be merged?

**Comment by RussKie** [comment]
Can you please share a new screenshot of how it looks?

**Comment by oriash93** [comment]
image

File: Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                {
+                    OnCheckBoxClicked(_checked);
+                    DataGridView.InvalidateCell(this);
+                }
+            }
+
+            base.OnMouseClick(e);
+        }
+
+        public void SetValue(bool value)
```
Comment (RussKie) [inline Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs]: 1. public API before protected please
2. rename to something like `SetChecked(bool)` or better yet refactor to `bool Checked`:

```
public bool Checked
{
    get { return _checked; }
    set 
    {
        if (_checked != value)
        {
            _checked = value;
            DataGridView.InvalidateCell(this);
        }
    }
}
```

File: Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            }
+
+            CheckBoxRenderer.DrawCheckBox(graphics, _checkBoxLocation, _cbState);
+        }
+
+        protected override void OnMouseClick(DataGridViewCellMouseEventArgs e)
+        {
+            Point p = new Point(e.X + _cellLocation.X, e.Y + _cellLocation.Y);
+            if (p.X >= _checkBoxLocation.X && p.X <= _checkBoxLocation.X + _checkBoxSize.Width
+            && p.Y >= _checkBoxLocation.Y && p.Y <= _checkBoxLocation.Y + _checkBoxSize.Height)
```
Comment (RussKie) [inline Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs]: misindented

File: Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
+﻿using System;
+using System.Drawing;
+using System.Windows.Forms;
+using System.Windows.Forms.VisualStyles;
+
+namespace DeleteUnusedBranches
+{
+    public delegate void CheckBoxClickedHandler(bool state);
```
Comment (RussKie) [inline Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs]: event handlers must follow https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/event

File: Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+    public class DataGridViewCheckBoxHeaderCellEventArgs : EventArgs
```
Comment (RussKie) [inline Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs]: one class per file

File: Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    internal class DataGridViewCheckBoxHeaderCell : DataGridViewColumnHeaderCell
+    {
+        private Point _checkBoxLocation;
+        private Size _checkBoxSize;
+        private bool _checked = false;
+        private Point _cellLocation = new Point();
+        private CheckBoxState _cbState = CheckBoxState.UncheckedNormal;
+        public event CheckBoxClickedHandler OnCheckBoxClicked;
```
Comment (RussKie) [inline Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs]: an empty line is missing

File: Plugins/DeleteUnusedBranches/DeleteUnusedBranchesForm.cs [hunk 6]
```diff
@@ -212,15 +215,29 @@ private void Refresh_Click(object sender, EventArgs e)
...
         }
 
+        private void CheckBoxHeader_OnCheckBoxClicked(bool state)
+        {
+            for (int i = 0; i < BranchesGrid.Rows.Count; i++)
+            {
+                DataGridViewRow row = BranchesGrid.Rows[i];
+                DataGridViewCheckBoxCell cell =
+                    (DataGridViewCheckBoxCell)row.Cells[nameof(deleteDataGridViewCheckBoxColumn)];
+                cell.Value = state;
```
Comment (RussKie) [inline Plugins/DeleteUnusedBranches/DeleteUnusedBranchesForm.cs]: follow the .NET event handlers design guidelines, and the column would be accessible via the `sender` argument

File: Plugins/DeleteUnusedBranches/DeleteUnusedBranchesForm.cs [hunk 7]
```diff
@@ -212,15 +215,29 @@ private void Refresh_Click(object sender, EventArgs e)
...
+            }
+
+            BranchesGrid.EndEdit();
```
Comment (RussKie) [inline Plugins/DeleteUnusedBranches/DeleteUnusedBranchesForm.cs]: why is this needed and where is `BeginInit` call?

**Comment by oriash93** [comment]
I need an advice, maybe any of you can help me.
I just found out there is **already** a `DataGridViewCheckBoxHeaderCell` class implemented under the `GitUI.CommandsDialogs` namespace. From the code, it is used in the **Verify Database** form, when clicking **Repository** > **Git maintenance** > **Recover lost objects**, though I can't see it visually:

image

Does anyone know how it was implemented or how it should be used?

File: Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs (hunk 1 above)
Comment (oriash93) [reply Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs]: Done.

File: Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs (hunk 2 above)
Comment (oriash93) [reply Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs]: Done.

File: Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs (hunk 3 above)
Comment (oriash93) [reply Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs]: Thanks for the reference, I learned something new. Done.

File: Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs (hunk 4 above)
Comment (oriash93) [reply Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs]: Done.  Also renamed it to `CheckBoxHeaderCellEventArgs`

File: Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs (hunk 5 above)
Comment (oriash93) [reply Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs]: Done.

File: Plugins/DeleteUnusedBranches/DeleteUnusedBranchesForm.cs (hunk 7 above)
Comment (oriash93) [reply Plugins/DeleteUnusedBranches/DeleteUnusedBranchesForm.cs]: It won't work otherwise. Added `CommitEdit(...)` call before.

**Comment by oriash93** [comment]
I need an advice, maybe any of you can help me.
I just found out there is **already** a `DataGridViewCheckBoxHeaderCell` class implemented under the `GitUI.CommandsDialogs` namespace. From the code, it is used in the **Verify Database** form, when clicking **Repository** > **Git maintenance** > **Recover lost objects**, though I can't see it visually:

image

Does anyone know how it was implemented or how it should be used?

**Comment by RussKie** [comment]
> I need an advice, maybe any of you can help me.
I just found out there is already a DataGridViewCheckBoxHeaderCell class implemented under the GitUI.CommandsDialogs namespace. From the code, it is used in the Verify Database form, when clicking Repository > Git maintenance > Recover lost objects, though I can't see it visually:

Good spot. The implementations look somewhat similar (in GH web UI), the existing one seems to update the checked state based on the state of all checkboxes in the column (`UpdateCheckedState`). This is perhaps something you may wish to copy.
Other than that, feel free to replace the existing implementation with yours (in a separate PR).

File: Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        private Point _cellLocation = new Point();
+        private CheckBoxState _cbState = CheckBoxState.UncheckedNormal;
+        private bool _checked = false;
+
+        public bool Checked
+        {
+            get { return _checked; }
+            set
+            {
+                _checked = value;
```
Comment (RussKie) [inline Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs]: you only want to do this if `_checked != value`, otherwise you will be invalidating the grid, which may be expensive.
https://github.com/gitextensions/gitextensions/pull/5100#discussion_r197610468

File: Plugins/DeleteUnusedBranches/CheckBoxHeaderCellEventArgs.cs [hunk 9]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+
+namespace DeleteUnusedBranches
+{
+    public class CheckBoxHeaderCellEventArgs : EventArgs
+    {
+        public bool Checked { get; set; }
```
Comment (RussKie) [inline Plugins/DeleteUnusedBranches/CheckBoxHeaderCellEventArgs.cs]: `EventArgs` payload (with few exceptions) must be readonly, otherwise you can't guarantee stability and consistency.

```
public class CheckBoxHeaderCellEventArgs : EventArgs
{
    public CheckBoxHeaderCellEventArgs(bool checked)
    {
        Checked = checked;
    }

    public bool Checked { get; }
}

File: Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+using System.Windows.Forms;
+using System.Windows.Forms.VisualStyles;
+
+namespace DeleteUnusedBranches
+{
+    public class DataGridViewCheckBoxHeaderCell : DataGridViewColumnHeaderCell
+    {
+        private Point _checkBoxLocation;
+        private Size _checkBoxSize;
+        private Point _cellLocation = new Point();
```
Comment (RussKie) [inline Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs]: no need to initialise, `Point` is a `struct`

File: Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        private CheckBoxState _cbState = CheckBoxState.UncheckedNormal;
+        private bool _checked = false;
```
Comment (RussKie) [inline Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs]: no need to initialise, `false` is the default value

File: Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+using System.Windows.Forms.VisualStyles;
+
+namespace DeleteUnusedBranches
+{
+    public class DataGridViewCheckBoxHeaderCell : DataGridViewColumnHeaderCell
+    {
+        private Point _checkBoxLocation;
+        private Size _checkBoxSize;
+        private Point _cellLocation = new Point();
+        private CheckBoxState _cbState = CheckBoxState.UncheckedNormal;
```
Comment (RussKie) [inline Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs]: no need to initialise, it is being set before use on line 57.
in fact I don't see any other uses of it, any reason it is not a local variable?

File: Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+using System.Drawing;
+using System.Windows.Forms;
+using System.Windows.Forms.VisualStyles;
+
+namespace DeleteUnusedBranches
+{
+    public class DataGridViewCheckBoxHeaderCell : DataGridViewColumnHeaderCell
+    {
+        private Point _checkBoxLocation;
+        private Size _checkBoxSize;
```
Comment (RussKie) [inline Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs]: any reason it is not a local variable?

**Comment by vbjay** [comment]
Adjusting what @russKie said. Eventargs that should not be changed by the
subscriber should be readonly.  So think of do you want the event
subscribers to change the value?

Here is an example where you would not make it readonly.
https://msdn.microsoft.com/en-us/library/system.componentmodel.canceleventargs.cancel(v=vs.110).aspx

File: Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs (hunk 13 above)
Comment (oriash93) [reply Plugins/DeleteUnusedBranches/DataGridViewCheckBoxHeaderCell.cs]: Yes, it is calculated inside the `Paint` method and used inside the `OnMouseClick` method.

**Comment by RussKie** [comment]
I've tried running your branch and I'm not seeing the checkbox in the heading
image

**Comment by RussKie** [comment]
Apologies, I was looking at a wrong branch

**Comment by RussKie** [comment]
Looks good.
A minor defect - the heading checkbox is not checked when the form is launched despite all checkboxes being checked.

**Comment by RussKie** [comment]
Please squash it and it will be good to go (once the heading is fixed)

**Comment by oriash93** [comment]
>Please squash it and it will be good to go (once the heading is fixed)

@RussKie Squash it all to one commit? Wouldn't it be easier to merge?

**Comment by drewnoakes** [comment]
>> Please squash it and it will be good to go (once the heading is fixed)

> @RussKie Squash it all to one commit?

I can't speak for RussKie but I believe the intention here is to go back through your PR and squash commits to create a cleaner history. You had an initial set of commits, then as fixes were made more commits were added. Squashing in such cases keeps the history cleaner, and it gives the impression you coded it perfectly the first time :)

> Wouldn't it be easier to merge?

Maybe you already knew about squashing and you meant here that it would be easier to squash during the merge:

image

In general it's better if the author squashes, as during that process you can combine/review the commit message.

**Comment by RussKie** [comment]
@drewnoakes will this create any issues for you, if merged?

**Comment by drewnoakes** [comment]
> will this create any issues for you, if merged?

No it looks good. I'll merge now and sort out any conflicts on the v3-ux branch.
